diff --git a/src/share/native/sun/misc/URLClassPath.c b/src/share/native/sun/misc/URLClassPath.c
--- a/src/share/native/sun/misc/URLClassPath.c
+++ b/src/share/native/sun/misc/URLClassPath.c
@@ -31,6 +31,10 @@
 #include "jvm.h"
 #include "jdk_util.h"
 
+/* defined in libverify.so/verify.dll (src file common/check_format.c) */
+extern jboolean VerifyClassname(char *utf_name, jboolean arrayAllowed);
+extern jboolean VerifyFixClassname(char *utf_name);
+
 #include "sun_misc_URLClassPath.h"
 
 extern char*
